### PR TITLE
feat: Prover interface and L1 interface crates

### PR DESCRIPTION
**Pull request by popzxc**
tl;dr:
- Removes crypto stuff from `zksync_types` (this was a primary goal for this PR)
- Reduces public API surface for general-purpose crates
- Introduces `l1_contract_interface` crate for stuff related to interacting with our main L1 contract.
- Introduces `prover_interface` crate as a point of interaction between prover and core subsystems.

❕Note: this is not a complete refactoring, there is still a lot of stuff that can be moved to either of these two crates, this will be worked on iteratively. This PR is meant to be just an improvement over what we currently have.

Rationale: `l1_contract_interface` and `prover_interface` are currently scattered across several crates (`zksync_types`, most notably) while actually being used only by few. Encapsulating them both reduces the public API surface plus reduces the number of (transitive) dependencies, including getting rid of some heavyweight stuff.

### CODE REVIEW

File: core/lib/l1_contract_interface/src/i_executor/methods/commit_batches.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+use zksync_types::{commitment::L1BatchWithMetadata, ethabi::Token};
```
[the remark this answers is not in the document]
Comment (popzxc) [reply core/lib/l1_contract_interface/src/i_executor/methods/commit_batches.rs:1]: I wanted to keep the naming consistent with the name of the contract data comes from.
I've had multiple occasions where I've wondered where in contracts this or that thing was defined.
Contracts (especially interfaces) change rarely, so this is unlikely to become outdated.

I'm open to opinions though.

**Comment by popzxc** [comment]
We've discussed with @EmilLuta that some of the feedback will be addressed in follow-up  PRs, and some will be tackled by him during the prover segregation project.

**Review by RomanBrodetski — APPROVED** [review]
